Add mapToMultiple mapping operation

peter-si opened the pull request:
Add polymorphic mapping opperation from https://github.com/mark-gerarts/automapper-plus/issues/64

Commit:
Add mapToMultiple mapping operation

peter-si commented:
Hi @mark-gerarts just a friendly reminder since you [mentioned](https://github.com/mark-gerarts/automapper-plus/pulls#issuecomment-783187120) you will take a look

mark-gerarts commented on src/MappingOperation/Implementations/MapToMultiple.php:
```diff
@@ -0,0 +1,107 @@
+
+/**
+ * Class MapTo.
+ * Allows a property to be mapped itself.
```
This docblock seems incorrect

peter-si replied:
right, copy&paste :) I'll remove it

mark-gerarts commented on src/MappingOperation/Implementations/MapToMultiple.php:
```diff
@@ -0,0 +1,107 @@
+    /**
+     * MapTo constructor.
+     * @param string[] $destinationClassList
+     * @param array $context
```
The phpdoc is slightly off here

mark-gerarts commented on src/MappingOperation/Implementations/MapToMultiple.php:
```diff
@@ -0,0 +1,107 @@
+     */
+    public function __construct(
+        array $destinationClassList,
+        array $context = []
```
Is there a special use case for this separate context?

peter-si replied:
I'm not using anything from it. I just added it so it is the same as on your other methods. Also I'm passing it to `map` method which I'm using `$returnValue = $this->mapper->map($value, $destinationClass, $context);`

mark-gerarts commented on src/MappingOperation/Implementations/MapToMultiple.php:
```diff
@@ -0,0 +1,107 @@
+    use ContextAwareTrait;
+
+    /**
+     * @var string
```
Should be `string[]` I believe?

mark-gerarts commented on README.md:
```diff
@@ -297,6 +298,18 @@ If you need to have the automapper available in your operation, you can
 implement the `MapperAwareInterface`, and use the `MapperAwareTrait`. The
 default `MapTo` and `MapFrom` operations use these.
 
+#### Dealing with polymorphic properties
```
It might be clearer if the section comes after the `mapTo` section.

peter-si replied:
I'm not sure I understand what you mean. Which `mapTo` section? it is after `mapTo` in the table

| MapTo | Maps the property to another class. Allows for [nested mappings](#dealing-with-nested-mappings). Supports both single values and collections. |
| MapToMultiple | Maps the property to list of another classes (maps to first found match). This can be used for [polymorphic properties](#dealing-with-polymorphic-properties). Allows for [nested mappings](#dealing-with-nested-mappings). Supports both single values and collections. |

mark-gerarts commented on src/MappingOperation/Implementations/MapToMultiple.php:
```diff
@@ -0,0 +1,107 @@
+ * Allows a property to be mapped itself.
+ * @package AutoMapperPlus\MappingOperation\Implementations
+ */
+class MapToMultiple extends DefaultMappingOperation implements
```
It seems you didn't include a `mapToMulitiple` method in the `Operation` class; feel free to do so.

peter-si replied:
added

mark-gerarts commented on src/MappingOperation/Implementations/MapToMultiple.php:
```diff
@@ -0,0 +1,107 @@
+        $returnValue = null;
+        $mappingFailed = false;
+        foreach ($this->destinationClassList as $destinationClass) {
+            if ($this->isCollection($value) === false) {
```
`if (!$this->isCollection($value))` is clearer imo

peter-si replied:
I prefer long notation, because you can see it better than just `!`, but I've change it according your suggestion

mark-gerarts commented on src/MappingOperation/Implementations/MapToMultiple.php:
```diff
@@ -0,0 +1,107 @@
+                    $mappingFailed = true;
+                }
+            } else {
+                foreach ($value as $item) {
```
It is not guaranteed the source property is iterable - if the user makes a mistake in their mapping, this would fatal out. It's better to explicitly check for `is_iterable` and throw a specific exception if it is not.

mark-gerarts replied:
Disregard this - I'm getting a bit too tired it seems :slightly_smiling_face:

mark-gerarts commented on src/MappingOperation/Implementations/MapToMultiple.php:
```diff
@@ -0,0 +1,107 @@
+     */
+    private function isCollection($variable): bool
+    {
+        return is_array($variable) || $variable instanceof Traversable;
```
Since we target PHP 7.1+ nowadays we can just use `is_iterable` instead (I should update `MapTo` as well).

mark-gerarts commented on test/MappingOperation/Implementations/MapToMultipleTest.php:
```diff
@@ -0,0 +1,139 @@
+        $mapToAnyOf->setMapper(AutoMapper::initialize(function (AutoMapperConfigInterface $config) {
+            $config->registerMapping(PolymorphicChildA::class, PolymorphicDtoA::class);
+            $config->registerMapping(PolymorphicChildB::class, PolymorphicDtoB::class);
+        }
```
`}));` on a single line looks cleaner imo - it might be another todo for me to add a linter to the codebase.

mark-gerarts commented on test/MappingOperation/Implementations/MapToMultipleTest.php:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Ensure the operation uses the assigned name resolver. See #17.
+     */
+    public function testItUsesTheNameResolver()
```
Thank you for including lots of tests!

mark-gerarts commented:
I'm sorry for the delay. I've done some peer review on the code - looking good, just a few small remarks! The only thing left for me to do is cloning the repo and playing around with the PR, doing some manual testing. Once I've done that and once you've had a chance to take a look at the comments, I'll be more than happy to merge this.

Thanks for your work!

Commit:
Add mapToMultiple mapping operation

peter-si commented:
Hi, sorry it took me so long, but I've finally updated it. Most of it should be fixed (I just didn't know where you want that Readme part:). I'm not used to these Github PR (I'm a Gitlab user) and I'm not sure if you see how the code changed regarding your comments (I noticed that outdated flag, but it doesn't look enough for me). Anyway take a look at it, hopefully it will be ok ;)

Also is there anything needed to be changed in [AutoMapperPlusBundle](https://github.com/mark-gerarts/automapper-plus-bundle) in order to get these changes? I think it should work out of the box after update? But maybe some configs? :man_shrugging:

peter-si commented:
Hi, @mark-gerarts have you had time to check this out?

fd6130 commented:
Still waiting this PR news.

mark-gerarts committed:
Minor refactoring tweaks

- README changes
- Extract some stuff into separate methods
- Add an extra test

mark-gerarts committed:
Rename MapToMultiple to MapToAnyOf

mark-gerarts commented:
@fd6130 thanks for the reminder, some of these things have been gathering dust for too long.

@peter-si sorry for getting back to you so late. I refactored some minor things, but most importantly, I renamed the feature from `MapToMultiple` to `MapToAnyOf`. I feel this is a more correct naming because of several reasons:

- `MapToMultiple` sounds more like it deals with mapping to a collection of things, even though it can be a single property as well.
- `MapToOneOf` as suggested in the related issue was disregarded there as well, since this implies the entire collection is of one type.
- `MapToAnyOf`is the name of this PR's branch and is used here and there in the tests, and sounds right to me.

I'll merge this into master, but will wait a bit before tagging a release after I've done some spring cleaning on open issues etc (poke me if this takes too long and I'll tag a new release regardless).

mark-gerarts commented:
...and forgot to answer this:

> Also is there anything needed to be changed in [AutoMapperPlusBundle](https://github.com/mark-gerarts/automapper-plus-bundle) in order to get these changes? I think it should work out of the box after update? But maybe some configs?

Yes, this will work out-of-the-box after a `composer update mark-gerarts/automapper-plus-bundle --with-all-dependencies` or equivalent.

Thanks again for the PR @peter-si!
